flow/src/main/java/com/sequenceiq/flow/core/Flow2Handler.java

```diff
@@ -451,7 +451,10 @@ public FlowIdentifier retryLastFailedFlow(Long resourceId, java.util.function.Co
     public FlowIdentifier retryLastFailedFlowFromStart(Long resourceId) {
         List<FlowLog> flowLogs = flowLogService.findAllForLastFlowIdByResourceIdOrderByCreatedDesc(resourceId);
         checkRetryable(flowLogs);
-        FlowLog firstSuccessfulStateLog = FlowLogUtil.getFirstStateLog(flowLogs);
+        FlowLog firstSuccessfulStateLog = FlowLogUtil.getFirstStateLog(flowLogs).orElseThrow(() -> {
+            LOGGER.info("Retry cannot be performed, because there is no latest flow: {}", resourceId);
+            return new BadRequestException("Retry cannot be performed, because there is no recent action.");
```

Comment:
same here, maybe it is worth to create a method in FlowLogUtil similar to this and use it everywhere:
```
public static FlowLog getFirstStateLogOrThrowInternalError(List<FlowLog> flowLogs, String crn) {
        return getFirstStateLog(flowLogs).orElseThrow(() -> {
            LOGGER.info("Retry cannot be performed, because there is no latest flow for resource: {}", crn);
            return new InternalServerErrorException("Retry cannot be performed, because there is no recent action.");
        });
    }
```

Reply:
of course you can use any other identifier instead of crn, like name or id